Add instructions for k8s node drain operation. (#343)
arangodb/kube-arangodb: docs/Manual/Deployment/Kubernetes/Drain.md

For maintenance work in k8s it is sometimes necessary to drain a k8s node,
which means removing all pods from it. Kubernetes offers a standard API
for this and our operator supports this - to the best of its ability.

Draining nodes is easy enough for stateless services, which can simply be
re-launched on any other node. However, for a stateful service this
operation is more difficult, and as a consequence more costly and there
are certain risks involved, if the operation is not done carefully
enough. To put it simply, the operator must first move all the data
stored on the node (which could be in a locally attached disk) to
another machine, before it can shut down the pod gracefully. Moving data
takes time, and even after the move, the distributed system ArangoDB has
to recover from this change, for example by ensuring data synchronicity
between the replicas in their new location.

{% hint 'warning' %}
Therefore, a systematic drain of all k8s nodes in sequence has to follow
a careful procedure, in particular to ensure that ArangoDB is ready to
move to the next step. This is necessary to avoid catastrophic data
loss, and is simply the price one pays for running a stateful service.

> {% endhint %}

Reviewer: @sleto-it I added a red hint box at the top, so maybe we can remove your warning here?

      "Endpoint": "ssl://my-arangodb-cluster-dbserver-wbsq47rz.my-arangodb-cluster-int.default.svc:8529",
      "SyncTime": "2019-02-21T09:14:24Z",
      "Version": "3.4.2-1",
      "Host": "my-arangodb-cluster-dbserver-wbsq47rz.my-arangodb-cluster-int.default.svc",
      "Timestamp": "2019-02-21T09:14:24Z",
      "Status": "GOOD",
      "SyncStatus": "SERVING",
      "Engine": "rocksdb",
      "ShortName": "DBServer0006",
      "Role": "DBServer",
      "CanBeDeleted": false
    },
    "AGNT-wrqmwpuw": {
      "Endpoint": "ssl://my-arangodb-cluster-agent-wrqmwpuw.my-arangodb-cluster-int.default.svc:8529",
      "Role": "Agent",
      "CanBeDeleted": false,
      "Version": "3.4.2-1",
      "Engine": "rocksdb",
      "Leader": "AGNT-oqohp3od",
      "Status": "GOOD",
      "LastAckedTime": 0.312
    },
    ... [some more entries, one for each instance]
  },
  "ClusterId": "210a0536-fd28-46de-b77f-e8882d6d7078",
  "error": false,
  "code": 200
}
```

Check that each instance has a `Status` field with the value `"GOOD"`.
Here is a shell command which makes this check easy, using the
[`jq` JSON pretty printer](https://stedolan.github.io/jq/):

```bash

> curl -k https://arangodb.9hoeffer.de:8529/_admin/cluster/health --user root: | jq . | grep '"Status"' | grep -v '"GOOD"'

Reviewer: @neunhoef Does jq support something similar to grep, so that we can make this a cross-platform command?

        "replicationFactor": 2,
        "shardKeys": [
          "_key"
        ],
        "shardingStrategy": "hash",
        "shards": {
          "s2010066": [
            "PRMR-vzeebvwf",
            "PRMR-e6hbjob1"
          ],
          "s2010062": [
            "PRMR-e6hbjob1",
            "PRMR-vzeebvwf"
          ],
          "s2010065": [
            "PRMR-e6hbjob1",
            "PRMR-vzeebvwf"
          ],
          "s2010067": [
            "PRMR-vzeebvwf",
            "PRMR-e6hbjob1"
          ],
          "s2010064": [
            "PRMR-vzeebvwf",
            "PRMR-e6hbjob1"
          ],
          "s2010063": [
            "PRMR-e6hbjob1",
            "PRMR-vzeebvwf"
          ]
        },
        "status": 3,
        "type": 2,
        "waitForSync": false
      },
      "indexes": [],
      "planVersion": 132,
      "isReady": true,
      "allInSync": true
    },
    ... [more collections following]
  ],
  "views": [],
  "tick": "38139421",
  "state": "unused"
}
```

Check that for all collections the attribute `"allInSync"` has
the value `true`. Note that it is necessary to do this for all databases!

Here is a shell command which makes this check easy:

```bash

> curl -k https://arangodb.9hoeffer.de:8529/_db/_system/_api/replication/clusterInventory --user root: | jq . | grep '"allInSync"' | sort | uniq -c

Reviewer: Does jq support this with a filter? There is group_by()...

considerably. Here is why:

If this step is not performed, we must choose
the grace period long enough to avoid any risk, as explained in the
previous section. However, this has a disadvantage which has nothing to
do with ArangoDB: We have observed, that some k8s internal services like
`fluentd` and some DNS services will always wait for the full grace
period to finish a node drain. Therefore, the node drain operation will
always take as long as the grace period. Since we have to choose this
grace period long enough for ArangoDB to move all data on the _DBServer_
pod away to some other node, this can take a considerable amount of
time, depending on the size of the data you keep in ArangoDB.

Therefore it is more time-efficient to perform the clean-out operation
beforehand. One can observe completion and as soon as it is completed
successfully, we can then issue the drain command with a relatively
small grace period and still have a nearly risk-free procedure.

To clean out a _DBServer_ manually, we have to use this API:

```
POST /_admin/cluster/cleanOutServer
```

… and send as body a JSON document like this:

```JSON
{"server":"DBServer0006"}
```

> (please compare the above output of the `/_admin/cluster/health` API).

Reviewer: @neunhoef  compare the cleanOutServer output to the cluster health output?

The value of the `"server"` attribute should be the name of the DBserver

> which is one the pod which shall be drained next. This uses the UI short

Reviewer: @neunhoef like this?
```suggestion
which is one of the pods which shall be drained next. This uses the UI short
```

name, alternatively one can use the internal name, which corresponds to

> the pod name: In our example, the pod name is:

Reviewer: ```suggestion
the pod name. In our example, the pod name is:
```

have been done successfully, it is safe to perform the drain
operation, similar to this command:

```bash
kubectl drain gke-draintest-default-pool-394fe601-glts --delete-local-data --ignore-daemonsets --grace-period=300
```

As described above, the options `--delete-local-data` for ArangoDB and
`--ignore-daemonsets` for other services have been added. A `--grace-period` of
300 seconds has been chosen because for this example we are confident that all the data on our _DBServer_ pod
can be moved to a different server within 5 minutes. Note that this is
**not saying** that 300 seconds will always be enough, regardless of how
much data is stored in the pod, your mileage may vary, moving a terabyte
of data can take considerably longer!

If the optional step of
[cleaning out a DBserver manually](#clean-out-a-dbserver-manually-optional)
has been performed beforehand, the grace period can easily be reduced to 60
seconds - at least from the perspective of ArangoDB, since the server is already
cleaned out, so it can be dropped readily and there is still no risk.

At the same time, this guarantees now that the drain is completed

> approximately within a minute.

Reviewer: Sounds like cleaning out the server manually is what we should recommend, yet we say it's optional. What are the other options? The problem I see with any grace period is that you never know if it's too long or too short, and that the draining process never finish before the end of the grace period...

## Things to check after a node drain

After a node has been drained, there will usually be one of the
_DBservers_ gone from the cluster. As a replacement, another _DBServer_ has
been deployed on a different node, if there is a different node
available. If not, the replacement can only be deployed when the
maintenance work on the drained node has been completed and it is

> uncordoned again. In this latter case, one should wait until the node is

Reviewer: is `uncordoned` a real word?

back up and the replacement pod has been deployed there.

> After that, one should perform the same checks as described in Section 

Reviewer: ```suggestion
After that, one should perform the same checks as described in
```

[things to check before a node drain](#things-to-check-in-arangodb-before-a-node-drain)
above.

Finally, it is likely that the shard distribution in the "new" cluster
is not balanced out. In particular, the new _DBSserver_ is not automatically
used to store shards. We recommend to
[re-balance](../../Administration/Cluster/#movingrebalancing-shards) the shard distribution,

> either manually by moving shards or by using the "Rebalance Shards"

Reviewer: ```suggestion
either manually by moving shards or by using the _Rebalance Shards_
```

> button in the "NODES/Shards" tab in the UI. This redistribution can take

Reviewer: ```suggestion
button in the _Shards_ tab under _NODES_ in the web UI. This redistribution can take
```
